### PR TITLE
[llvm-debuginfo-analyzer] Common handling of unsigned attribute values.

### PR DESCRIPTION
**Pull request by CarlosAlbertoEnciso**
- In the DWARF reader, for those attributes that can have an unsigned value, allow for the following cases:
  * Is an implicit constant
  * Is an optional value
- The testing is done by creating a file with generated DWARF, using `DwarfGenerator` (generate DWARF debug info for unit tests).

### CODE REVIEW

**Review by pogo59 — COMMENTED** [review]
The test makes it pretty clear this is functionally the correct thing to do. But it covers only the inline attribute; are we missing tests for the other affected attributes? (I can't remember whether there is a unittest for the reader, but if there is, that would seem like a good place to cover all these cases.)

**Comment by CarlosAlbertoEnciso** [comment]
We have unittest for the reader: `llvm/unittests/DebugInfo/LogicalView`. But we don't have the ability to create logical elements with specific DWARF tags/attributes.
I will look at adding some generic basic support using `llvm/unittests/DebugInfo/DWARF`

**Comment by CarlosAlbertoEnciso** [comment]
@pogo59 Using the existing `llvm/unittests/DebugInfo/DWARF/DwarfGenerator` module, added a unittest to include the affected attributes.

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+using namespace llvm;
+using namespace llvm::logicalview;
+using namespace llvm::dwarf;
+using namespace llvm::dwarf::utils;
+
+namespace {
+
+// Helper function to get the first compile unit.
+LVScopeCompileUnit *getFirstCompileUnit(LVScopeRoot *Root) {
+  EXPECT_NE(Root, nullptr);
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:38]:
```suggestion
  ASSERT_NE(Root, nullptr);
```
Use the `ASSERT_*` macros when the test will later crash if the condition is not true. This is especially the case when verifying a pointer is non-null and will be dereferenced.

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+  EXPECT_EQ(TypeImplicitRangeDieDG.getTag(), DW_TAG_subrange_type);
+
+  // Save the generated DWARF file to disk.
+  EXPECT_TRUE(DG->saveFile(Path));
+}
+
+// Check the logical elements basic properties.
+void checkElementAttributes(LVReader *Reader) {
+  LVScopeRoot *Root = Reader->getScopesRoot();
+  EXPECT_NE(Root, nullptr);
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:218]: This is redundant with a check inside getFirstCompileUnit

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+using namespace llvm::dwarf;
+using namespace llvm::dwarf::utils;
+
+namespace {
+
+// Helper function to get the first compile unit.
+LVScopeCompileUnit *getFirstCompileUnit(LVScopeRoot *Root) {
+  EXPECT_NE(Root, nullptr);
+  const LVScopes *CompileUnits = Root->getScopes();
+  EXPECT_NE(CompileUnits, nullptr);
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:40]:
```suggestion
  ASSERT_NE(CompileUnits, nullptr);
```

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+
+  const LVScopes *Scopes = CompileUnit->getScopes();
+  ASSERT_NE(Scopes, nullptr);
+  ASSERT_EQ(Scopes->size(), 3u);
+
+  // Check values.
+  LVScopes::const_iterator ScopeIter = Scopes->begin();
+  EXPECT_NE(ScopeIter, nullptr);
+  LVScope *Scope = static_cast<LVScope *>(*ScopeIter);
+  EXPECT_NE(Scope, nullptr);
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:229]:
```suggestion
  ASSERT_NE(Scope, nullptr);
```

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+  EXPECT_EQ(Scope->getInlineCode(), 2);        // Element
+  EXPECT_EQ(Scope->getVirtualityCode(), 3);    // Element
+  EXPECT_EQ(Scope->getCallFilenameIndex(), 5); // ScopeFunctionInlined
+  EXPECT_EQ(Scope->getCallLineNumber(), 5);    // ScopeFunctionInlined
+  EXPECT_EQ(Scope->getFilenameIndex(), 7);     // Element
+  EXPECT_EQ(Scope->getLineNumber(), 7);        // Element
+  EXPECT_EQ(Scope->getDiscriminator(), 8);     // ScopeFunctionInlined
+
+  // Check no-values.
+  EXPECT_NE(++ScopeIter, nullptr);
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:240]: Another iteration check

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+  Scope = static_cast<LVScope *>(*ScopeIter);
+  EXPECT_NE(Scope, nullptr);
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:242]:
```suggestion
  ASSERT_NE(Scope, nullptr);
```

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+  EXPECT_EQ(Scope->getCallFilenameIndex(), 1); // ScopeFunctionInlined
+  EXPECT_EQ(Scope->getCallLineNumber(), 0);    // ScopeFunctionInlined
+  EXPECT_EQ(Scope->getFilenameIndex(), 1);     // Element
+  EXPECT_EQ(Scope->getLineNumber(), 0);        // Element
+  EXPECT_EQ(Scope->getDiscriminator(), 0);     // ScopeFunctionInlined
+
+  // Check implicit values.
+  EXPECT_NE(++ScopeIter, nullptr);
+  Scope = static_cast<LVScope *>(*ScopeIter);
+  EXPECT_NE(Scope, nullptr);
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:255]:
```suggestion
  ASSERT_NE(Scope, nullptr);
```

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+
+// Helper function to get the first compile unit.
+LVScopeCompileUnit *getFirstCompileUnit(LVScopeRoot *Root) {
+  EXPECT_NE(Root, nullptr);
+  const LVScopes *CompileUnits = Root->getScopes();
+  EXPECT_NE(CompileUnits, nullptr);
+  EXPECT_EQ(CompileUnits->size(), 1u);
+
+  LVScopes::const_iterator Iter = CompileUnits->begin();
+  EXPECT_NE(Iter, nullptr);
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:44]: Iterators generally should be compared to `end()` ? Also this should be an `ASSERT_NE`.

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+  EXPECT_NE(Root, nullptr);
+  LVScopeCompileUnit *CompileUnit = getFirstCompileUnit(Root);
+
+  const LVScopes *Scopes = CompileUnit->getScopes();
+  ASSERT_NE(Scopes, nullptr);
+  ASSERT_EQ(Scopes->size(), 3u);
+
+  // Check values.
+  LVScopes::const_iterator ScopeIter = Scopes->begin();
+  EXPECT_NE(ScopeIter, nullptr);
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:227]: Iteration should be compared to `end()` ? Also, use `ASSERT_NE` here.

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+  EXPECT_EQ(Scope->getInlineCode(), 0);        // Element
+  EXPECT_EQ(Scope->getVirtualityCode(), 0);    // Element
+  EXPECT_EQ(Scope->getCallFilenameIndex(), 1); // ScopeFunctionInlined
+  EXPECT_EQ(Scope->getCallLineNumber(), 0);    // ScopeFunctionInlined
+  EXPECT_EQ(Scope->getFilenameIndex(), 1);     // Element
+  EXPECT_EQ(Scope->getLineNumber(), 0);        // Element
+  EXPECT_EQ(Scope->getDiscriminator(), 0);     // ScopeFunctionInlined
+
+  // Check implicit values.
+  EXPECT_NE(++ScopeIter, nullptr);
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:253]: Another iterator check

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+  EXPECT_EQ(Scope->getLineNumber(), 7);        // Element
+  EXPECT_EQ(Scope->getDiscriminator(), 8);     // ScopeFunctionInlined
+
+  const LVSymbols *Symbols = CompileUnit->getSymbols();
+  ASSERT_NE(Symbols, nullptr);
+  ASSERT_EQ(Symbols->size(), 3u);
+
+  // Check values.
+  LVSymbols::const_iterator SymbolIter = Symbols->begin();
+  EXPECT_NE(SymbolIter, nullptr);
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:271]:
```suggestion
  ASSERT_NE(SymbolIter, nullptr);
```

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+  LVSymbol *Symbol = static_cast<LVSymbol *>(*SymbolIter);
+  EXPECT_NE(Symbol, nullptr);
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:273]:
```suggestion
  ASSERT_NE(Symbol, nullptr);
```

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+  EXPECT_EQ(Symbol->getBitSize(), 1); // Symbol
+
+  // Check no-values.
+  EXPECT_NE(++SymbolIter, nullptr);
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:277]: This pattern continues, I won't add comments on all of them.

File: llvm/unittests/DebugInfo/LogicalView/CMakeLists.txt [hunk 14]
```diff
@@ -1,17 +1,22 @@
...
+  AsmPrinter
   DebugInfoLogicalView
   MC
   MCDisassembler
   TargetParser
   )
 
 add_llvm_unittest_with_input_files(DebugInfoLogicalViewTests
+  ../DWARF/DwarfGenerator.cpp
+  ../DWARF/DwarfUtils.cpp
```
Comment (pogo59) [inline llvm/unittests/DebugInfo/LogicalView/CMakeLists.txt:15]: This is okay for now, but really these components should be moved into their own library so they can be shared between unittests.

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp (hunk 13 above)
Comment (CarlosAlbertoEnciso) [reply llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp:277]: Removed the repeated comments.

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp (hunk 1 above)
Comment (CarlosAlbertoEnciso) [reply llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp]: Changed all `EXPECT_NE` to `ASSERT_NE`.

**Note**: There is a restriction on using `EXPECT_NE` inside a non-void function
https://chromium.googlesource.com/external/github.com/google/googletest/+/HEAD/docs/advanced.md#assertion-placement

For the non-void functions: `getFirstCompileUnit` and `createReader`, I used standard if checks and the caller will use `ASSERT_NE` on the returned value.

File: llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp (hunk 3 above)
Comment (CarlosAlbertoEnciso) [reply llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp]: The check has been removed from getFirstCompileUnit. See previous comment about assertion-placement.

File: llvm/unittests/DebugInfo/LogicalView/CMakeLists.txt (hunk 14 above)
Comment (CarlosAlbertoEnciso) [reply llvm/unittests/DebugInfo/LogicalView/CMakeLists.txt:15]: That is a good point. I will create a PR to record that request.

**Review by pogo59 — APPROVED** [review]
LGTM

**Comment by CarlosAlbertoEnciso** [comment]
@pogo59 Thanks very much for your valuable review.

**Comment by CarlosAlbertoEnciso** [comment]
Looking at the reported failures on those specific builders:
- Link errors: Missing explicitly references to `DebugInfoDWARF` and `Object` in the `CMakeList.txt`
- Compile errors: error: comparison of integers of different signs:

File: llvm/unittests/DebugInfo/LogicalView/CMakeLists.txt [hunk 15]
```diff
@@ -1,17 +1,22 @@
 set(LLVM_LINK_COMPONENTS
+  ${LLVM_TARGETS_TO_BUILD}
```
Comment (nico) [inline llvm/unittests/DebugInfo/LogicalView/CMakeLists.txt:2]: Doesn't this make the next 3 lines superfluous?

Comment (CarlosAlbertoEnciso) [reply llvm/unittests/DebugInfo/LogicalView/CMakeLists.txt:2]: @nico That is a good point. With your suggested change, all pre-commit test pass.
Opened https://github.com/llvm/llvm-project/pull/118052

**Pull request merged**